### ISSUE TITLE
Niche optimization in JsValue

### ISSUE THREAD

**Issue by WorldSEnder** [rustwasm/wasm-bindgen#3201]
### Motivation

I have a bunch of code using `Option<SomeJsBindgenType>`, e.g. `Option<JsValue>`, `Option<Node>`. These types are represented as a `JsValue` under-the-hood, which is declared as

```
pub struct JsValue {
    idx: u32,
    _marker: marker::PhantomData<*mut u8>, // not at all threadsafe
}
```

Would it be possible to include a niche-optimization here, e.g. by using `idx: NonZeroU32`, so that the above Option types have size 4 instead of size 8? It seems some indices are already reserved for values in "stack space", so could we free one additional index to serve as a niche for optimizing `Option<JsValue>`?

### Additional Context

[NonZeroU32](https://doc.rust-lang.org/std/num/struct.NonZeroU32.html#)

**Comment by Liamolucko** [comment]
Yeah, it should be pretty easy to reduce the stack size by 1 and make index 0 into a niche. I think this makes sense to do.
